torch/csrc/jit/passes/create_functional_graphs.cpp

```diff
@@ -0,0 +1,233 @@
+
+    // Functional Graphs are not responsible for maintaining aliasing
+    // relationships. If an output of a functional graph escapes scope
+    // or is mutated then we might change semantics of the program if
+    // aliasing relationships are changed.
+    // For now, we don't allow any values which are mutated into the functional
+    // graph, and we don't allow any nodes which have outputs that escape scope.
+    // Possible Future Improvements:
+    // - allow inputs to have mutations so long as there are no mutations in the
+    // graph
+    // - allow functional graphs to have at most one value that can escape scope
+    // - allow outputs which alias the wildcard set but do not "re-escape"
+
+    for (Value* v : n->outputs()) {
+      bool has_writers = aliasDb_->hasWriters(v);
+      bool escapes_scope = aliasDb_->escapesScope(v);
+      if (has_writers) {
+        mutated_values_.insert(v);
+      }
+      functional_node = functional_node && !escapes_scope && !has_writers;
+    }
+
+    for (Block* block : n->blocks()) {
+      auto functional_block = AnalyzeFunctionalSubset(block);
+      functional_node = functional_node && functional_block;
+    }
+
+    auto inputs = n->inputs();
+    functional_node = functional_node &&
+        std::all_of(inputs.begin(), inputs.end(), [&](Value* v) {
+                        return !mutated_values_.count(v);
+                      });
+    if (functional_node) {
+      functional_nodes_.insert(n);
+    }
+    return functional_node;
+  }
+
+  void AnalyzeFunctionalSubset(at::ArrayRef<Block*> blocks) {
+    for (Block* block : blocks) {
+      AnalyzeFunctionalSubset(block);
+    }
+  }
+
+  bool AnalyzeFunctionalSubset(Block* block) {
+    bool is_functional_block = true;
+    // block inputs will not yet have been iterated through,
+    // so we need to add them to our set of mutated & escape values.
+    for (Value* v : block->inputs()) {
+      bool has_writers = aliasDb_->hasWriters(v);
+      if (has_writers) {
+        mutated_values_.insert(v);
+      }
+    }
+    for (Node* n : block->nodes()) {
+      bool functional = AnalyzeFunctionalSubset(n);
+      is_functional_block = is_functional_block && functional;
+    }
+    return is_functional_block;
+  }
+
+  std::unordered_set<Node*> functional_nodes_;
+  std::unordered_set<Value*> mutated_values_;
+  std::shared_ptr<Graph> graph_;
+  std::unique_ptr<AliasDb> aliasDb_ = nullptr;
+  size_t minSubgraphSize_ = 6;
+};
+
+void InlineFunctionalGraphs(Block* block) {
+  for (auto it = block->nodes().begin(); it != block->nodes().end();) {
+    Node* n = *it;
+    it++;
+    for (Block* b : n->blocks()) {
+      InlineFunctionalGraphs(b);
+    }
+    if (n->kind() == prim::FunctionalGraph) {
+      SubgraphUtils::unmergeSubgraph(n);
+    }
+  }
+}
+
+} // namespace
+
+void CreateFunctionalGraphs(const std::shared_ptr<Graph>& graph) {
+  // Run Constant Pooling so constants get hoisted
+  ConstantPooling(graph);
```

Comment:
We can leave this in, but as a general principle I want to start enforcing that graph passes don't run preprocessing and can ingest any sort of graph. It's not super important in this case, but I feel like it'll force us to make fewer assumptions which makes the passes more composable.